File: PopularMovies/src/main/java/com/mchapagai/movies/api/impl/SearchAPIImpl.java
package com.mchapagai.movies.api.impl;

import com.mchapagai.core.response.movies.MovieListResponse;
import com.mchapagai.core.response.people.CombinedPersonResponse;
import com.mchapagai.core.response.people.PersonResponse;
import com.mchapagai.movies.api.SearchAPI;
import com.mchapagai.movies.service.MovieSearchService;

import javax.inject.Provider;

import io.reactivex.Observable;
import io.reactivex.Single;

public class SearchAPIImpl implements SearchAPI {

    private final Provider<MovieSearchService> movieService;

    public SearchAPIImpl(Provider<MovieSearchService> service) {
        this.movieService = service;
    }

    @Override
    public Single<PersonResponse> getPersonDetailsById(int personId) {
        return movieService.get().getPersonDetailsById(personId);
    }

    @Override
    public Observable<CombinedPersonResponse> getPersonCombinedDetailsById(int personId) {
        return movieService.get().getPersonCombinedDetailsById(personId);
    }

    @Override
    public Observable<MovieListResponse> searchMovies(String query) {
        return movieService.get().searchMovies(query);
    }
}

File: Core/src/main/java/com/mchapagai/core/api/SearchAPIImpl.kt
package com.mchapagai.core.api

import com.mchapagai.core.response.search.SearchResponse
import com.mchapagai.core.service.SearchService
import com.mchapagai.core.utils.RxUtils
import io.reactivex.Observable

class SearchAPIImpl(private val service: SearchService) : SearchAPI {
    override fun search(query: String): Observable<SearchResponse> {
        return service.search(query)
            .compose(RxUtils.applyObservableSchedulers())
    }
}